content/backend/graphql-node-typescript/3-graphql-server.md

```diff
@@ -0,0 +1,242 @@
+
+<Instruction>
+
+Start by adding `graphql-helix` to your project:
+
+```bash
+npm install --save graphql-helix
+```
+
+</Instruction>
+
+Now, in order to accept incoming GraphQL requests and handle them, you can use GraphQL-Helix. You'll start by adding an endpoint for `POST /graphql` for our server, and then use `graphql-helix` to normalize the incoming request, and then execute it against the GraphQL schema:
+
+<Instruction>
+
+Change `src/index.ts` and replace the dummy request handler with the following:
+
+```typescript{3,4,9-32}(path="hackernews-node-ts/src/index.ts")
+import 'graphql-import-node';
+import fastify from "fastify";
+import { getGraphQLParameters, processRequest, Request } from "graphql-helix";
+import { schema } from "./schema";
+
+async function main() {
+  const server = fastify();
+
+  server.route({
+    method: "POST",
+    url: "/graphql",
+    handler: async (req, reply) => {
+      const request: Request = {
+        headers: req.headers,
+        method: req.method,
+        query: req.query,
+        body: req.body,
+      };
+
+      const { operationName, query, variables } = getGraphQLParameters(request);
+
+      const result = await processRequest({
+        request,
+        schema,
+        operationName,
+        query,
+        variables,
+      });
+
+      if (result.type === "RESPONSE") {
+        reply.send(result.payload);
+      } else {
+        reply.send({ error: "Stream not supported at the moment" });
+      }
+    }
+  });
+
+  server.listen(3000, "0.0.0.0", () => {
+    console.log(`Server is running on http://localhost:3000/`);
+  });
+}
+
+main();
+```
+
+</Instruction>
+
+You can now try to run your server, and you should be able to run our query in a HTTP-based GraphQL request. 
+
+<Instruction>
+
+If you have `curl` installed on your machine, you should be able to run and see the GraphQL response (you can also use any kind of HTTP requests tools, like [Postman](https://www.postman.com/)):
+
+```bash
+curl -X POST http://localhost:3000/graphql -H "Content-type: application/json" --data-raw '{"query": "query { info }"}'
+```
+</Instruction>
+
+So let's break down the code snippet for the server:
+
+1. Your created an new HTTP request handler, of type `POST`, based on the `/graphql` path. 
+1. The handler takes the incoming raw HTTP request and builds a `Request` object out of it.
+1. Then, the `getGraphQLParameters` method of `graphql-helix` extracts all the neccesary information from the raw request. 
+1. The `processRequest` gets all the input parameters, and runs the GraphQL request pipeline for us.
+1. The result of the GraphQL operation is being returned and sent to the client.  
+
+> GraphQL-Helix supports real-time GraphQL and stream-based responses, that's why your code need to check it's result before sending it. At the moment, our server only supports regular responses (`type="RESPONSE"`). You'll later add support for more transports.  
```

Comment:
This sentence introduces a lot of concepts (stream-based responses, transports) that a beginner is unlikely to be familiar with. Please simply this sentence, explain the necessary concepts and provide appropriate resources where necessary.

```suggestion
> GraphQL-Helix supports real-time GraphQL and stream-based responses, that's why your code needs to check the result before sending it. At the moment, our server only supports regular responses (`type="RESPONSE"`). You'll later add support for more transports.  
```